corehq/apps/sso/tests/test_async_handlers.py

```diff
@@ -63,12 +79,70 @@ def _get_post_data(self, object_name=None):
         }
 
 
+class TestAsyncHandlerSecurity(BaseAsyncHandlerTest):
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.second_account = generator.get_billing_account_for_idp()
+
+    def setUp(self):
+        super().setUp()
+        self.request = _get_request(self.second_account)
+        self.request_superuser = _get_request(self.second_account, is_superuser=True)
+
+    def test_get_linked_objects_response_throws_404(self):
+        self.request.POST = self._get_post_data()
+        handler = IdentityProviderAdminAsyncHandler(self.request)
+        with self.assertRaises(Http404):
+            handler.get_linked_objects_response
+
+    def test_add_object_response_throws_404(self):
+        self.request.POST = self._get_post_data('vaultwax.nl')
+        handler = IdentityProviderAdminAsyncHandler(self.request)
+        with self.assertRaises(Http404):
+            handler.add_object_response
+
+    def test_remove_object_response_throws_404(self):
+        self.request.POST = self._get_post_data('vaultwax.nl')
+        handler = IdentityProviderAdminAsyncHandler(self.request)
+        with self.assertRaises(Http404):
+            handler.remove_object_response
+
+    def test_get_linked_objects_response_does_not_throw_404_for_superusers(self):
+        self.request_superuser.POST = self._get_post_data()
+        handler = IdentityProviderAdminAsyncHandler(self.request_superuser)
+        self.assertEqual(
+            handler.get_linked_objects_response,
+            {'linkedObjects': []}
+        )
+
+    def test_add_object_response_does_not_throw_404_for_superusers(self):
+        self.request_superuser.POST = self._get_post_data('vaultwax.nl')
+        handler = IdentityProviderAdminAsyncHandler(self.request_superuser)
+        self.assertEqual(
+            handler.add_object_response,
+            {'linkedObjects': ['vaultwax.nl']}
+        )
+
+    def test_remove_object_response_does_not_throw_404_for_superusers(self):
+        AuthenticatedEmailDomain.objects.create(
+            identity_provider=self.idp,
+            email_domain='vaultwax.nl'
+        )
+        self.request_superuser.POST = self._get_post_data('vaultwax.nl')
+        handler = IdentityProviderAdminAsyncHandler(self.request_superuser)
+        self.assertEqual(
+            handler.remove_object_response,
+            {'linkedObjects': []}
+        )
+
+
 class TestIdentityProviderAdminAsyncHandler(BaseAsyncHandlerTest):
 
     def setUp(self):
         super().setUp()
-        self.request = RequestFactory().get('/sso/test')
-        self.request.method = 'POST'
+        self.request = _get_request(self.idp.owner)
 
     def tearDown(self):
         AuthenticatedEmailDomain.objects.all().delete()
```

Comment:
I know this didn't change here, but deleting SQL objects on teardown is not necessary (and just makes the tests slower) since they will be cleaned up automatically on TestCase transaction rollback.

There are several more of these unnecessary teardown methods throughout this file.